### DISCUSSION TITLE
company-yasnippet doesn't play nice with other completion backends

### DISCUSSION THREAD

**Discussion by sethen** [company-mode/company-mode#1249] [Q&A]
The only way I can get `company-yansippet` to play nicely is like so:

```elisp
(set (make-local-variable 'company-backends) '((:separate company-yasnippet company-capf) company-keywords)))
```

If I turn the code to something like this:

```elisp
(set (make-local-variable 'company-backends) '(company-yasnippet company-capf company-keywords))
```

It will only complete the snippets and not move on to the `company-capf` backend... it won't even show those completions.  What's going on here?  How can I get all of these backends to play nice with each other and show _all_ completions at once?

**Comment by dgutov** [comment]
To use several backends together, instead of "first one which can take responsibility", you need to *group* them.

In the example above, it means

```el
(set (make-local-variable 'company-backends) '((company-yasnippet company-capf company-keywords)))
```

**Reply by sethen** [reply]
This issue is the way that the completions are sorted using each piece of code.

With this piece of code:

```elisp
(set (make-local-variable 'company-backends) '(company-yasnippet company-capf company-keywords))
```

I get this:
![Screenshot from 2021-10-24 20-34-41](https://user-images.githubusercontent.com/1493530/138619230-a029d5ef-0825-4916-8431-1d0584d93119.png)

Which is what I want.  However when I use this piece of code:

```elisp
(set (make-local-variable 'company-backends) '((company-yasnippet company-capf company-keywords)))
```

I get this:
![Screenshot from 2021-10-24 20-33-36](https://user-images.githubusercontent.com/1493530/138619253-63e9696f-7d8c-4dbc-ba16-3e82f04b411b.png)

Which I don't want.  I am unsure why changing this code messes with the order in which it appears now?  It started sorted using substrings rather than literal completions?

**Reply by dgutov** [reply]
I actually don't understand how you got that result on the screenshot #1. Shouldn't you get only yasnippet snippets instead?

**Reply by dgutov** [reply]
But if you add the keyword `:separate` to the group, it will keep the sorting order as returned by the backends.

**Reply by sethen** [reply]
I don't understand really either lol -- disabling orderless gave me exactly what I wanted tho

**Reply by sethen** [reply]
@dgutov Actually upon further inspection trying to use multiple company backends together is not working for me at all in regards to sorting.  No matter what I try I get one dominant completion backend and nothing else.  This has nothing to do with orderless like I thought.

**Reply by dgutov** [reply]
Try

```el
(setq-local company-backends '((:separate company-capf company-yasnippet company-keywords)))
```

It will keep the sorting order for completions coming from every backend, and keep them in the order that the backends are specified in, too.
